### DISCUSSION TITLE
Mapping large items to model

### DISCUSSION THREAD

**Discussion by ParadiseFallen** [mobxjs/mobx-state-tree#1968] [Q&A]
Hi. In my app i has api service that ferch some data for me.
I has 
```ts
const Book = types.model({
  id: types.numIdentifier,
  title: types.string,
});

const BookStore = types.mpdel({
  items : types.array(Book)
})
.actions((self)=>({
  loadItems : flow(function* (){
    const api = getEnv(self).services.bookService;
    const response = yield bookService.loadAll();
    
    // how to set values and keep em as much performant as possible?
  })
}));
```

**Comment by ParadiseFallen** [comment]
much issue also is `How to work with DTO`.

When api return normalized data and requires normalazed.
And snapshots is not what api consumes?

**Comment by coolsoftwaretyler** [comment]
Hey @ParadiseFallen - I have a similar use case in my app. We have a pretty large payload from an internal API and load it into MST. We have an array of API data that needs to end up as an array of MST nodes.

We've gone through two different iterations:

1. At first, our data didn't match the MST models very well, so we pretty much had to loop through the array from the API, make some small transformations, and then create individual objects and add them to the MST array.
2. In our second iteration, we were able to get the API to return data in a shape that matches our MST model. The API and state agree more closely on the typings, so we can pretty easily just load them directly. In your example, this could look like:

```js
const Book = types.model({
  id: types.numIdentifier,
  title: types.string,
});

const BookStore = types.mpdel({
  items : types.array(Book)
})
.actions((self)=>({
  loadItems : flow(function* (){
    const api = getEnv(self).services.bookService;
    const response = yield bookService.loadAll();
    const booksFromAPI = yield response.json() // Or whatever gets you an array of your objects
    
    self.items = [...booksFromAPI]
  })
}));
```

I can't prove that this is the *most* performant way to do this, but we load thousands of domain-specific objects in one operation like that, and it happens *quite fast* in a React Native app. I don't expect this would be much of a performance problem, but if you have some performance monitoring that tells you it's a problem, I'd love to hear more and see if I can help you with a better solution.
